contracts/libs/0.8.x/MerkleLib.sol

```diff
@@ -0,0 +1,101 @@
+// SPDX-License-Identifier: LGPL-3.0-only
+// Created By: Art Blocks Inc.
+
+import "@openzeppelin-4.7/contracts/utils/cryptography/MerkleProof.sol";
+
+pragma solidity ^0.8.0;
+
+/**
+ * @title Art Blocks Merkle Library
+ * @notice [TO FILL OUT]
+ * @author Art Blocks Inc.
+ */
+
+library MerkleLib {
+    using MerkleProof for bytes32[];
+    uint256 public constant DEFAULT_MAX_INVOCATIONS_PER_ADDRESS = 1;
+    bytes32 public constant CONFIG_MERKLE_ROOT = "merkleRoot";
+    bytes32 public constant CONFIG_USE_MAX_INVOCATIONS_PER_ADDRESS_OVERRIDE =
+        "useMaxMintsPerAddrOverride"; // shortened to fit in 32 bytes
+    bytes32 public constant CONFIG_MAX_INVOCATIONS_OVERRIDE =
+        "maxMintsPerAddrOverride"; // shortened to match format of previous key
+
+    struct ProjectConfig {
+        bool maxHasBeenInvoked;
+        bool priceIsConfigured;
+        // initial value is false, so by default, projects limit allowlisted
+        // addresses to a mint qty of `DEFAULT_MAX_INVOCATIONS_PER_ADDRESS`
+        bool useMaxInvocationsPerAddressOverride;
+        // a value of 0 means no limit
+        // (only used if `useMaxInvocationsPerAddressOverride` is true)
+        uint24 maxInvocationsPerAddressOverride;
+        uint24 maxInvocations;
+        uint256 pricePerTokenInWei;
+    }
+
+    function updateMerkleRoot(
+        mapping(address => mapping(uint256 => bytes32))
+            storage projectMerkleRoot,
+        uint256 _projectId,
+        address _coreContract,
+        bytes32 _root
+    ) internal {
+        require(_root != bytes32(0), "Root must be provided");
+        projectMerkleRoot[_coreContract][_projectId] = _root;
+    }
+
+    function hashAddress(address _address) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(_address));
+    }
+
+    function processProofForAddress(
+        bytes32[] calldata _proof,
+        address _address
+    ) internal pure returns (bytes32) {
+        return _proof.processProofCalldata(hashAddress(_address));
+    }
+
+    function verifyAddress(
+        bytes32 _proofRoot,
+        bytes32[] calldata _proof,
+        address _address
+    ) internal pure returns (bool) {
+        return _proof.verifyCalldata(_proofRoot, hashAddress(_address));
+    }
+
+    function projectMaxInvocationsPerAddress(
+        uint256 _projectId,
+        address _coreContract,
+        mapping(address => mapping(uint256 => ProjectConfig))
+            storage projectConfig
+    ) public view returns (uint256) {
```

Comment:
I think we probably want this to be internal, given that we intend to use this as a bundled internal library ya?